### PR TITLE
Make Tuple mutable.

### PR DESCRIPTION
**Pull request by ericbottard**
This fixes spring-cloud/spring-cloud-stream#290
This resolves spring-cloud/spring-cloud-stream-modules#166

### CODE REVIEW

**Comment by markpollack** [comment]
How about we keep an immutable option and in addition provide a new mutable one.

**Comment by ericbottard** [comment]
Yep, thought about it, as an option on the builder. We'd want the default to be mutable though (or create a new mime type)

**Comment by markpollack** [comment]
We should err on the side of keeping backwards comparability with XD to minimize change when users convert to the new spring cloud stream model.

**Comment by ericbottard** [comment]
So are you advocating for a new mime-type for example? And when users want a mutable tuple, they'd explicitly ask for it?

**Comment by mbogoevici** [comment]
How exactly can this affect backwards compatibility though - the old contract remains unchanged.

**Comment by mminella** [comment]
+1 for allowing for an immutable option in some form.

**Comment by ericbottard** [comment]
I added a way to create immutable Tuples on the builder

File: spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultMutableTuple.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * @author Eric Bottard
+ */
+public class DefaultMutableTuple extends DefaultTuple implements MutableTuple {
+
+	public DefaultMutableTuple(List<String> names, List<Object> values, ConfigurableConversionService configurableConversionService) {
+		super(names, values, configurableConversionService);
+	}
+
+	@Override
+	public synchronized void setValue(int index, Object value) {
```
Comment (pperalta) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultMutableTuple.java]: Unless reads are also `synchronized`, I'm not sure that making these methods `synchronized` is worthwhile. IMO, either the entire class needs to be made thread safe, or the JavaDoc should explicitly state that it isn't thread safe.

Comment (ericbottard) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultMutableTuple.java]: very good point

**Comment by markpollack** [comment]
You are right Marius, it would not impact existing code, but it would break a rather important contract going forward.  The analogy in SI would be MutableMessage and Message.  I had though that perhaps one day this would also be a nice replacement to FieldSet in batch - since it was created before the conversion service.  The current set of tests - and in some cases behavior - is mostly derived from the FieldSet tests in batch.

File: spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultMutableTuple.java (hunk 1 above)
Comment (markpollack) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultMutableTuple.java]: I'd go for avoiding putting synchronized on it and stating it in the docs.  There could always be another sync'd variant created by the builder.

**Comment by ericbottard** [comment]
Removed synchronized, Forced use of TupleBuilder (see also https://github.com/spring-cloud/spring-cloud-stream-modules/pull/168) and documented threadsafety

File: spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultTuple.java [hunk 2]
```diff
@@ -36,22 +36,23 @@
...
+	protected List<Object> values;
 
-	private transient ConfigurableConversionService configurableConversionService;
+	protected transient ConfigurableConversionService configurableConversionService;
 
-	private transient Converter<Tuple, String> tupleToStringConverter = new TupleToJsonStringConverter();
+	protected transient Converter<Tuple, String> tupleToStringConverter = new TupleToJsonStringConverter();
 
-	public DefaultTuple(List<String> names, List<Object> values, ConfigurableConversionService 
+	/*package*/ DefaultTuple(List<String> names, List<Object> values, ConfigurableConversionService
```
Comment (markpollack) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultTuple.java]: Why change to package level protection?  Was thinking the hard enforcement of not having a ctor vs. only builder was a bit too restrictive.

File: spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/JsonStringToTupleConverter.java [hunk 3]
```diff
@@ -38,6 +38,7 @@ public JsonStringToTupleConverter() {
 
 	@Override
 	public Tuple convert(String source) {
+		System.out.println("Converting " + source);
```
Comment (markpollack) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/JsonStringToTupleConverter.java]: No system.out...

File: spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/JsonStringToTupleConverter.java [hunk 4]
```diff
@@ -70,7 +71,9 @@ else if (node.isNumber()) {
 		catch (Exception e) {
 			throw new RuntimeException(e);
 		}
-		return builder.build();
+		Tuple result = builder.build();
```
Comment (markpollack) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/JsonStringToTupleConverter.java]: Should just be `return builder.build()`

File: spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultMutableTuple.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Extension of {@link DefaultTuple} that also allows tuple {@link MutableTuple mutation}.
+ *
+ * <p>Note that this implementation is not threadsafe.</p>
+ *
+ * @author Eric Bottard
+ */
+public class DefaultMutableTuple extends DefaultTuple implements MutableTuple {
+
+	/*package*/ DefaultMutableTuple(List<String> names, List<Object> values, ConfigurableConversionService configurableConversionService) {
```
Comment (markpollack) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/DefaultMutableTuple.java]: Similar comment as with protected ctor for DefaultTuple

File: spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/TupleBuilder.java [hunk 6]
```diff
@@ -68,8 +70,22 @@
...
+	public static TupleBuilder immutableTuple() {
+		return new TupleBuilder(false);
+	}
+
+	/**
+	 * Return a new builder that will create {@link MutableTuple}.
+	 */
 	public static TupleBuilder tuple() {
-		return new TupleBuilder();
+		return new TupleBuilder(true);
```
Comment (markpollack) [inline spring-cloud-stream-tuple/src/main/java/org/springframework/cloud/stream/tuple/TupleBuilder.java]: The default should be immutable, as it was before, with mutable being the new option.

**Comment by markpollack** [comment]
I made changes and merged.  TubleBuilder.tuple() returns an immutable tuple like before.  I also made the ctors public, mostly concerned about backward compat, but we can discuss more.  Fixed the other minor issues.  

See 7fb834981ab6e987d1d0c784aebfbe642677e9a9

**Comment by ericbottard** [comment]
We absolutely need a new mime type for mutable tuple then. One of the aim of this PR is to use mutable tuples with the pmml module (via `inputType=`)

**Comment by markpollack** [comment]
This opens up another discussion, Mark F. mention some changes to how type conversion is done.  He can explain more.  We haven't had any perf complaints wrt to using immutable objects, so given that I'd suggest that we defer the goal of pmml using mutable tuples until we can plan out this change better.

**Comment by markfisher** [comment]
My comment about type-conversion is that we should move away from representing Java types as custom MIME types altogether, but instead explicitly recognize the distinction between conversion to/from the *wire* vs. conversion before/after message handling (i.e. it should be a 2-step process: could be `text/plain` or `application/octet-stream` on the wire and either of those then convert to `Tuple` or `String` or `Foo`).

**Pull request closed without merging**